File: ts-src/session/list.ts
import { readdirSync } from "node:fs";
import { join } from "node:path";
import { getProjectSessionsDir } from "../config/paths.ts";
import { loadSessionMeta } from "./jsonl.ts";

export interface SessionInfo {
	id: string;
	name: string | undefined;
	created: string;
	model: string;
	cwd: string;
	messageCount: number;
	firstUserMessage: string | undefined;
	forkedFrom: string | undefined;
}

/**
 * List all sessions in the given (or default) sessions directory.
 * Parses each .jsonl file's session_meta, counts messages, extracts first user message.
 * Returns sorted by created descending.
 */
export async function listSessions(sessionDir?: string): Promise<SessionInfo[]> {
	const dir = sessionDir ?? getProjectSessionsDir();

	let files: string[];
	try {
		files = readdirSync(dir).filter((f) => f.endsWith(".jsonl"));
	} catch {
		return [];
	}

	const sessions: SessionInfo[] = [];

	for (const file of files) {
		const filePath = join(dir, file);
		const meta = await loadSessionMeta(filePath);
		if (!meta) continue;

		const content = await Bun.file(filePath).text();
		const lines = content
			.trim()
			.split("\n")
			.filter((l) => l.trim())
			.map((l) => {
				try {
					return JSON.parse(l) as Record<string, unknown>;
				} catch {
					return null;
				}
			})
			.filter((obj): obj is Record<string, unknown> => obj !== null);

		const messages = lines.filter((obj) => "role" in obj);
		const firstUser = messages.find((obj) => obj.role === "user");
		let firstUserMessage: string | undefined;
		if (firstUser && typeof firstUser.content === "string") {
			firstUserMessage = firstUser.content.length > 100 ? firstUser.content.slice(0, 100) : firstUser.content;
		}

		// Look for session_name marker (last one wins)
		let name: string | undefined = meta.name as string | undefined;
		for (const line of lines) {
			if (line.type === "session_name" && typeof line.name === "string") {
				name = line.name;
			}
		}

		sessions.push({
			id: meta.id,
			name,
			created: meta.created,
			model: meta.model,
			cwd: meta.cwd,
			messageCount: messages.length,
			firstUserMessage,
			forkedFrom: (meta.forked_from as string) ?? undefined,
		});
	}

	sessions.sort((a, b) => b.created.localeCompare(a.created));
	return sessions;
}

/**
 * Find a session file by target. If target is empty/undefined, returns the most recent.
 * If target looks like a UUID (or prefix), match by id. Otherwise try name lookup.
 * Returns file path string or null.
 */
export async function findSession(target: string | undefined, sessionDir?: string): Promise<string | null> {
	const dir = sessionDir ?? getProjectSessionsDir();

	let files: string[];
	try {
		files = readdirSync(dir).filter((f) => f.endsWith(".jsonl"));
	} catch {
		return null;
	}

	// Load all metas
	const entries: { meta: { id: string; name?: string; created: string }; filePath: string }[] = [];
	for (const file of files) {
		const filePath = join(dir, file);
		const meta = await loadSessionMeta(filePath);
		if (!meta) continue;

		// Also scan for session_name markers
		let name: string | undefined = meta.name as string | undefined;
		const content = await Bun.file(filePath).text();
		const lines = content
			.trim()
			.split("\n")
			.filter((l) => l.trim());
		for (const line of lines) {
			try {
				const parsed = JSON.parse(line);
				if (parsed.type === "session_name" && typeof parsed.name === "string") {
					name = parsed.name;
				}
			} catch {}
		}

		entries.push({ meta: { id: meta.id, name, created: meta.created }, filePath });
	}

	if (!target || target.trim() === "") {
		// Return most recent
		if (entries.length === 0) return null;
		entries.sort((a, b) => b.meta.created.localeCompare(a.meta.created));
		return entries[0]?.filePath ?? null;
	}

	// Try UUID match (exact or prefix)
	const byId = entries.find((e) => e.meta.id === target || e.meta.id.startsWith(target));
	if (byId) return byId.filePath;

	// Try name match
	const byName = entries.find((e) => e.meta.name === target);
	if (byName) return byName.filePath;

	return null;
}

File: src/session/list.rs
//! List recent sessions and look one up by id/name/most-recent.

use std::path::{Path, PathBuf};

use serde_json::Value;

use super::jsonl::load_session_meta;
use crate::config::paths::get_project_sessions_dir;

#[derive(Debug, Clone)]
pub struct SessionInfo {
    pub id: String,
    pub name: Option<String>,
    pub created: String,
    pub model: String,
    pub cwd: String,
    pub message_count: usize,
    pub first_user_message: Option<String>,
    pub forked_from: Option<String>,
}

fn truncate_chars(s: &str, max_chars: usize) -> String {
    s.chars().take(max_chars).collect()
}

pub fn list_sessions(session_dir: Option<&Path>) -> Vec<SessionInfo> {
    let dir = match session_dir {
        Some(d) => d.to_path_buf(),
        None => get_project_sessions_dir(None),
    };
    let entries = match std::fs::read_dir(&dir) {
        Ok(e) => e,
        Err(_) => return Vec::new(),
    };
    let mut out = Vec::new();
    for entry in entries.flatten() {
        let path = entry.path();
        if path.extension().and_then(|e| e.to_str()) != Some("jsonl") {
            continue;
        }
        let meta = match load_session_meta(&path) {
            Some(m) => m,
            None => continue,
        };
        let content = match std::fs::read_to_string(&path) {
            Ok(c) => c,
            Err(_) => continue,
        };
        let parsed_lines: Vec<Value> = content
            .lines()
            .filter(|l| !l.trim().is_empty())
            .filter_map(|l| serde_json::from_str(l).ok())
            .collect();

        let messages: Vec<&Value> = parsed_lines
            .iter()
            .filter(|v| v.get("role").is_some())
            .collect();
        let first_user = messages
            .iter()
            .find(|m| m.get("role").and_then(Value::as_str) == Some("user"));
        let first_user_message = first_user.and_then(|m| {
            m.get("content").and_then(Value::as_str).map(|s| {
                if s.chars().count() > 100 {
                    truncate_chars(s, 100)
                } else {
                    s.to_string()
                }
            })
        });

        let mut name = meta.name.clone();
        for line in &parsed_lines {
            if line.get("type").and_then(Value::as_str) == Some("session_name") {
                if let Some(n) = line.get("name").and_then(Value::as_str) {
                    name = Some(n.to_string());
                }
            }
        }

        out.push(SessionInfo {
            id: meta.id.clone(),
            name,
            created: meta.created.clone(),
            model: meta.model.clone(),
            cwd: meta.cwd.clone(),
            message_count: messages.len(),
            first_user_message,
            forked_from: meta.forked_from.clone(),
        });
    }
    out.sort_by(|a, b| b.created.cmp(&a.created));
    out
}

pub fn find_session(target: Option<&str>, session_dir: Option<&Path>) -> Option<PathBuf> {
    let dir = match session_dir {
        Some(d) => d.to_path_buf(),
        None => get_project_sessions_dir(None),
    };
    let entries = std::fs::read_dir(&dir).ok()?;
    let mut entries_v: Vec<(SessionInfo, PathBuf)> = Vec::new();
    for entry in entries.flatten() {
        let path = entry.path();
        if path.extension().and_then(|e| e.to_str()) != Some("jsonl") {
            continue;
        }
        let meta = match load_session_meta(&path) {
            Some(m) => m,
            None => continue,
        };
        let mut name = meta.name.clone();
        if let Ok(content) = std::fs::read_to_string(&path) {
            for line in content.lines() {
                if let Ok(v) = serde_json::from_str::<Value>(line) {
                    if v.get("type").and_then(Value::as_str) == Some("session_name") {
                        if let Some(n) = v.get("name").and_then(Value::as_str) {
                            name = Some(n.to_string());
                        }
                    }
                }
            }
        }
        entries_v.push((
            SessionInfo {
                id: meta.id.clone(),
                name,
                created: meta.created.clone(),
                model: meta.model.clone(),
                cwd: meta.cwd.clone(),
                message_count: 0,
                first_user_message: None,
                forked_from: meta.forked_from.clone(),
            },
            path,
        ));
    }

    let target = target.map(str::trim).filter(|s| !s.is_empty());
    match target {
        None => {
            entries_v.sort_by(|a, b| b.0.created.cmp(&a.0.created));
            entries_v.into_iter().next().map(|(_, p)| p)
        }
        Some(t) => {
            if let Some((_, p)) = entries_v
                .iter()
                .find(|(i, _)| i.id == t || i.id.starts_with(t))
            {
                return Some(p.clone());
            }
            entries_v
                .into_iter()
                .find(|(i, _)| i.name.as_deref() == Some(t))
                .map(|(_, p)| p)
        }
    }
}
